src/core/symbol/MerkleHashBuilder.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 SYMBOL
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { sha3_256 } from 'js-sha3';
+
+export class MerkleHashBuilder {
+    /**
+     * The list of hashes used to calculate root hash.
+     *
+     * @member {Uint8Array}
+     */
+    private hashes: Uint8Array[] = new Array<Uint8Array>();
+
+    /**
+     * Calculates the merkle hash.
+     *
+     * @returns {Uint8Array} Root merkle hash
+     */
+    public final(): Uint8Array {
+        if (this.hashes.length === 0) {
+            return new Uint8Array(32);
+        }
+
+        let numRemainingHashes = this.hashes.length;
+        while (numRemainingHashes > 1) {
+            for (let i = 0; i < numRemainingHashes; i += 2) {
+                if (i + 1 < numRemainingHashes) {
+                    this.hashes.splice(i / 2, 0, this.hash([this.hashes[i], this.hashes[i + 1]]));
+                    continue;
+                }
+
+                // if there is an odd number of hashes, duplicate the last one
+                this.hashes.splice(i / 2, 0, this.hash([this.hashes[i], this.hashes[i]]));
+                ++numRemainingHashes;
+            }
+            numRemainingHashes /= 2;
```

Comment:
this should be integer division. i'm not sure if TS has that concept. if not, can use Math.floor.